tools/SourceKit/include/SourceKit/Support/FileSystemProvider.h

```diff
@@ -0,0 +1,59 @@
+//===--- FileSystemProvider.h - ---------------------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_SOURCEKIT_SUPPORT_FILESYSTEMPROVIDER_H
+#define LLVM_SOURCEKIT_SUPPORT_FILESYSTEMPROVIDER_H
+
+#include "llvm/ADT/IntrusiveRefCntPtr.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/Support/VirtualFileSystem.h"
+
+namespace SourceKit {
+
+/// Allows clients of SourceKit to specify custom llvm::vfs::FileSystems to be
+/// used while serving a request.
+///
+/// Requests to SourceKit select FileSystemProviders by specifying
+/// 'key.vfs.name', and pass arguments to the FileSystemProviders by
+/// specifying 'key.vfs.args'. SourceKit then passes the given arguments to the
+/// selected FileSystemProvider, and uses the resulting llvm::vfs::FileSystem
+/// while serving the request.
+///
+/// The following requests currently support custom FileSystemProviders (other
+/// requests respond with an invalid request error if you try):
+/// - source.request.editor.open: Associates the given custom filesystem with
+///                               this editor file, so that all subsequent
+///                               operations on this editor file use it. Sending
+///                               another 'source.request.editor.open' with
+///                               'key.vfs.name' replaces the filesystem.
+///                               Sending another 'source.request.editor.open'
+///                               without 'key.vfs.name' resets it to the default
+///                               filesystem.
```

Comment:
I don't think we need to document this behaviour for performing editor.open for a document that is already open.  That's not really a valid thing to do, and if we were starting from scratch I would make it an error.  Currently we log a warning.

Reply:
Oh interesting, thanks for the note.

I have been getting sema diagnostics by re-calling "editor.open" every time a file changes, waiting for the document update notification, and then calling "editorReplaceText" (with blank replacement) to read out the sema diagnostics. It seems to have been working. I guess that is not the right way to do this? What would the right way be?

Reply:
The system is designed for incrementally updating the contents with replacetext.  Is there a reason you can't use that?  If you need to re-open the document, you should close it before openining it again.